src/qmapshack/mouse/CMouseNormal.cpp

```diff
@@ -438,9 +442,16 @@ void CMouseNormal::slotSelectArea() const
 void CMouseNormal::showContextMenu(const QPoint &point)
 {
     QMenu menu(canvas);
-    if(curPOI.pos != NOPOINTF)
+    if(curPOIs.count() > 0 && curPOIs.count() <= 5)
+    {
+        for(int i = 0; i < curPOIs.count(); i++)
+        {
+            menu.addAction(QIcon("://icons/32x32/AddWpt.png"),  tr("Add POI %1 as Waypoint").arg(curPOIs[i].name), this, [this, i] {slotAddPoi(i);});
```

Comment:
Why not iterate over `curPOIs` and add the objects as copy to the lambda? 

Reasoning: An index is always a bit dangerous as the list might change between adding these actions and calling the slot. In the slot you are missing the range check for the index -> potential crash due to assertion. But checking the index will not save you from selecting a wrong element if the order did change or the list did change completely. 

Therefore the most fail safe method would be to copy the objects into the lambda.